collector/therm_darwin.go

```diff
@@ -0,0 +1,183 @@
+// Copyright 2021 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// +build !notherm
+
+package collector
+
+/*
+#cgo LDFLAGS: -framework IOKit -framework CoreFoundation
+#include <stdio.h>
+#include <CoreFoundation/CoreFoundation.h>
+#include <IOKit/IOKitLib.h>
+#include <IOKit/pwr_mgt/IOPMLib.h>
+#include <IOKit/pwr_mgt/IOPM.h>
+
+struct ref_with_ret {
+    CFDictionaryRef ref;
+    IOReturn ret;
+};
+
+struct ref_with_ret FetchThermal();
+
+struct ref_with_ret FetchThermal() {
+    CFDictionaryRef ref;
+    IOReturn ret;
+    ret = IOPMCopyCPUPowerStatus(&ref);
+    struct ref_with_ret result = {
+            ref,
+            ret,
+    };
+    return result;
+}
+*/
+import "C"
+
+import (
+	"errors"
+	"fmt"
+	"github.com/go-kit/kit/log"
+	"github.com/prometheus/client_golang/prometheus"
+	"unsafe"
+)
+
+type thermCollector struct {
+	cpuSchedulerLimit typedDesc
+	cpuAvailableCPU   typedDesc
+	cpuSpeedLimit     typedDesc
+	logger            log.Logger
+}
+
+const therm = "therm"
+
+func init() {
+	registerCollector("therm", defaultEnabled, NewThermCollector)
+}
+
+// NewThermCollector returns a new Collector exposing current CPU power levels.
+func NewThermCollector(logger log.Logger) (Collector, error) {
+	return &thermCollector{
+		cpuSchedulerLimit: typedDesc{
+			desc: prometheus.NewDesc(
+				prometheus.BuildFQName(namespace, therm, "cpu_scheduler_limit"),
+				"Represents the percentage (0-100) of CPU time available. 100% at normal operation. The OS may limit this time for a percentage less than 100%.",
+				nil,
+				nil),
+			valueType: prometheus.GaugeValue,
+		},
+		cpuAvailableCPU: typedDesc{
+			desc: prometheus.NewDesc(
+				prometheus.BuildFQName(namespace, therm, "cpu_available_cpu"),
+				"Reflects how many, if any, CPUs have been taken offline. Represented as an integer number of CPUs (0 - Max CPUs).",
```

Comment:
Shouldn't this be `"cpu_offline"`?

Reply:
Actually, the value of that is indeed available CPUs. I copy that description from https://opensource.apple.com/source/xnu/xnu-2422.100.13/iokit/IOKit/pwr_mgt/IOPM.h.auto.html, it might bring confusing.

Reply:
What should we do about that name thing? 🤔

Reply:
I think we should ignore the doc string in the code, as it's confusing.  How about this "How many CPUs are available and not taken offline, ratio of 0-1".

Reply:
cpu_available_cpu will show you the core number of CPU, on my machine with 6c12t, the value is 12.

Maybe not using ratio is better?

Reply:
Ahh, right, not a ratio.